### PR TITLE
[Eng] Github action to sync generation to future

### PR DESCRIPTION
**Pull request by VeryEarly**
## Description
NOTE: this action will be deleted once new process was available

future branch is based on main branch and had been migrate to future structure

This temporary github action:

- triggered by commit in generation branch
- checkout to future branch and cherry-pick trigger commit
- reset commit and autorest/build-module the changed autogen submodule
- commit everything under src and push to future branch

these above process imitates the actual PR/CI process in the future

## Mandatory Checklist

- Please choose the target release of Azure PowerShell. (⚠️**Target release** is a different concept from **API readiness**. Please click below links for details.)
  - [ ] [General release](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Public preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Private preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Engineering build](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] No need for a release

- [ ] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For SDK-based development mode, update `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`.
        * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. 
    * For autorest-based development mode, include the changelog in the PR description.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

**Review by YanaXu — APPROVED** [review]

**Comment by YanaXu** [comment]
The trigger is better to be schduled instead of on push for this case.

**Comment by isra-fel** [comment]
> The trigger is better to be schduled instead of on push for this case.

From resource consumption's perspective?

**Comment by VeryEarly** [comment]
> > The trigger is better to be schduled instead of on push for this case.
> 
> From resource consumption's perspective?

Discussed with Yan offline, this action does not require picking every commit precisely, so it's acceptable if few of them were missed because of frequent trigger.

File: .github/workflows/sync-generation-to-future.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        $repoRoot = '${{ github.workspace }}'
+
+        git fetch origin generation
+        git cherry-pick $commitId --strategy-option theirs
+        git reset HEAD~1
+        <#
+          calculate changed modules
+        #>
+        $modulePaths = @()
+        git status --short | Where-Object { $_ -match ".*src.*" } | ForEach-Object { $modulePaths += $_.substring($_.IndexOf("src"), $_.IndexOf(".Autorest")+6) }
```
Comment (isra-fel) [inline .github/workflows/sync-generation-to-future.yml:40]: although not frequent, there may be case or pr update "src/lib" or "src/shared"

File: .github/workflows/sync-generation-to-future.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        if (0 -eq $modulePaths.Count) {
+          return
+        }
+        $modulePaths | Select-Object -Unique
+        foreach ($modulePath in $modulePaths) {
```
Comment (isra-fel) [inline .github/workflows/sync-generation-to-future.yml:45]: $modulePaths is still not unique on line 45

File: .github/workflows/sync-generation-to-future.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        Install-Module -Name PowerShellGet -RequiredVersion 2.2.3 -Force;
+        Install-Module -Name platyPS -RequiredVersion 0.14.2 -Force;
+
+        git config user.email "65331932+azure-powershell-bot@users.noreply.github.com";
+        git config user.name "azure-powershell-bot";
+        $commitId = '${{ github.sha }}'
+        $repoRoot = '${{ github.workspace }}'
+
+        git fetch origin generation
+        git cherry-pick $commitId --strategy-option theirs
```
[the remark this answers is not in the document]
Comment (VeryEarly) [reply .github/workflows/sync-generation-to-future.yml:34]: This strategy is supposed to suppress most conflict, and future branch can be easily reset if there are serious conflict between it and generation branch

**Pull request merged**